components/support/jwcrypto/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Theorically, everything done in this crate could and should be done in a JWT library.
+//! However, none of the existing rust JWT libraries can handle ECDH-ES encryption, and API choices
+//! made by their authors make it difficult to add this feature.
+//! In the past, we chose cjose to do that job, but it added three C dependencies to build and link
+//! against: jansson, openssl and cjose itself.
+
+pub use error::JwCryptoError;
+use error::Result;
+use rc_crypto::agreement::EphemeralKeyPair;
+use serde_derive::{Deserialize, Serialize};
+use std::str::FromStr;
+
+pub mod ec;
+mod error;
+
+pub enum EncryptionParameters {
+    // ECDH-ES in Direct Key Agreement mode.
+    #[allow(non_camel_case_types)]
+    ECDH_ES {
+        enc: EncryptionAlgorithm,
+        peer_jwk: Jwk,
+    },
+}
+
+pub enum DecryptionParameters {
+    // ECDH-ES in Direct Key Agreement mode.
+    #[allow(non_camel_case_types)]
+    ECDH_ES { local_key_pair: EphemeralKeyPair },
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+enum Algorithm {
+    #[serde(rename = "ECDH-ES")]
+    #[allow(non_camel_case_types)]
+    ECDH_ES,
+}
+#[derive(Serialize, Deserialize, Debug)]
+pub enum EncryptionAlgorithm {
+    A256GCM,
+}
+
+impl EncryptionAlgorithm {
+    fn algorithm_id(&self) -> &'static str {
+        match self {
+            Self::A256GCM => "A256GCM",
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct JweHeader {
+    alg: Algorithm,
+    enc: EncryptionAlgorithm,
+    epk: Option<Jwk>,
+    apu: Option<String>,
+    apv: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(tag = "kty")]
+pub enum Jwk {
+    #[serde(alias = "ec")]
+    EC(ec::ECKeysParameters),
+}
+
+#[derive(Debug)]
+pub struct CompactJwe {
```

Comment:
It took a bit of work to make it usable everywhere, but making `CompactJwe` a type was worth the trouble.